File: backend/src/main/java/io/smarthome/core/automation/Rule.java
package io.smarthome.core.automation;

import io.smallrye.mutiny.Uni;

public interface Rule {

    Uni<Void> evaluate(RuleContext context);

    default boolean appliesTo(String eventType) {
        return true;
    }

    /** Runtime enablement (e.g. from configuration); combined with the @AutomationRule annotation flag. */
    default boolean isEnabled() {
        return true;
    }
}

File: backend-python/app/automation/rule.py
"""Mirror of the Rule interface and the @AutomationRule qualifier."""

from dataclasses import dataclass

from app.automation.context import RuleContext


@dataclass(frozen=True)
class RuleMetadata:
    name: str
    description: str = ""
    enabled: bool = True


class Rule:
    metadata: RuleMetadata

    def evaluate(self, context: RuleContext) -> None:
        raise NotImplementedError

    def applies_to(self, event_type: str) -> bool:
        return True

    def is_enabled(self) -> bool:
        """Runtime enablement (e.g. from configuration); combined with the
        metadata flag."""
        return True
